### PR TITLE
feat:  recall from hidden fields and attributes

### LINKED ISSUE #2353: [FEATURE] Prefill question text, description text, answer options and forward URLs with hidden field values (via URL params)
### Is your feature request related to a problem? Please describe.
Currently using recall informations inside a survey is only possible for previous questions. But users would also love to use this for hidden fields. 

### Describe the solution you'd like

Add hidden fields to the list of recall options inside the survey editor so hidden fields values can also be used for all cases where recall is activated.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Dhruwang**
## What does this PR do?

Fixes #2353 #2327 

- Now we also allow recalling from hidden fields and attributes
- Initially it was mainly focused on question, so we used namings like `recallQuestion`, `selectRecallQuestion` but now its more generic, so now we use `recallItem`, `selectRecallItems`
- Breaks single response card into smaller components 

Attributes

Screenshot 2024-05-10 at 12 02 34 PM

Hidden Fields 

Screenshot 2024-05-10 at 12 01 36 PM

Also made the recall object more expandable to support other fields in future

## How should this be tested?

For link surveys -> recall from questions and hidden fields is allowed
Fo app surveys -> recall from questions and attributes is allowed

### Required

- [x] Filled out the "How to test" section in this PR
- [x] Read [How we Code at Formbricks](<[https://github.com/formbricks/formbricks/blob/main/CONTRIBUTING.md](https://formbricks.com/docs/contributing/how-we-code)>)
- [x] Self-reviewed my own code
- [x] Commented on my code in hard-to-understand bits
- [x] Ran `pnpm build`
- [x] Checked for warnings, there are none
- [x] Removed all `console.logs`
- [x] Merged the latest changes from main onto my branch with `git pull origin main`
- [x] My changes don't cause any responsiveness issues
- [x] First PR at Formbricks? [Please sign the CLA!](https://cla-assistant.io/formbricks/formbricks) Without it we wont be able to merge it 🙏

### Appreciated

- [x] If a UI change was made: Added a screen recording or screenshots to this PR
- [x] Updated the Formbricks Docs if changes were necessary

### CODE REVIEW

**Comment by jobenjada** [comment]
hey @Dhruwang 

thanks a lot, love how you always crank out a bunch of cool improvements out on a given day :D 🚀 

Here's what I noticed:

1. If the @ is the first character in the input, I cannot select the option:

image

-------

2. We need to make clear that the second input field is the placeholder for the second recall item:

image

Easiest would be to add a placeholder text with the recall item label so the hiddenField id or the question.

-------

3. The recall of attributes does not seem to work for me. I've recorded a video to show how I tried:

https://github.com/formbricks/formbricks/assets/72809645/636b8dc9-e0b6-43e5-a87e-d48d3345f716

-------

4. Kombobox / Search bar for dropdown: some users have thousands of attributes. It would make sense to add the same combobox that we're now using in the Logic Editor:

see here: https://github.com/formbricks/formbricks/pull/2584

-------

Did you try your implementation with each question types? I normally find bugs when I do that 😉

-----

Other than that, it works great :) Thanks!

**Comment by Dhruwang** [comment]
Hey @jobenjada , have updated the PR ! 😊

Fallback input with placeholders 

Screenshot 2024-05-13 at 12 52 18 PM

Searchable dropdown

Screenshot 2024-05-13 at 12 21 25 PM


Also tested the recall from attributes and it seems to work fine 🤔

https://github.com/formbricks/formbricks/assets/67850763/57c8d8fa-ea2c-41be-9965-535590e129f2


In the video that you provided, did you miss the survey saving ?

**Comment by jobenjada** [comment]
Hey Dru!

Looks good :)  Here's my feedback:

1. Keyboard usability

With the search bar, we lost keyboard usability: 
image

Pls make sure that `Arrow down` lets people select one of the fields. Similarly, if they go back up with the keys they should hop back into the Search bar once they reach the top of the list.

-------

2. UI changes in response card

As laid out in Slack

------

Please ping me once you've reworked the attribute logic, then I'll test it.

Thanks!

**Review by jobenjada — APPROVED** [review]
looks great! :)

File: apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/components/SummaryPage.tsx [hunk 1]
```diff
@@ -105,8 +108,8 @@ export const SummaryPage = ({
   const searchParams = useSearchParams();
 
   survey = useMemo(() => {
```
Comment (pandeymangg) [inline apps/web/app/(app)/environments/[environmentId]/surveys/[surveyId]/(analysis)/summary/components/SummaryPage.tsx:110]: Please don't mutate the function parameter `survey`, you can make a new variable here `surveyMemoized` and use it in the rest of the places 😊 

```
const surveyMemoized = useMemo(() => {
    return checkForRecallInHeadline(survey, "default", attributeClasses);
  }, [survey, attributeClasses]);
```

**Review by pandeymangg — CHANGES_REQUESTED** [review]
Hey @Dhruwang thanks for the PR, looks great 🚀 
I just had one issue with it that I've written, pls take a look at it. And also make the changes to `recallToHeadline` function and the `ee` service as we discussed 😊

File: apps/web/app/api/v1/client/[environmentId]/app/sync/lib/utils.ts [hunk 2]
```diff
@@ -0,0 +1,81 @@
+import { parseRecallInfo } from "@formbricks/lib/utils/recall";
+import { TLegacySurvey } from "@formbricks/types/LegacySurvey";
+import { TAttributes } from "@formbricks/types/attributes";
+import { TSurvey } from "@formbricks/types/surveys";
+
+export const checkForAttributeRecall = (survey: TSurvey, attributes: TAttributes): TSurvey => {
```
Comment (mattinannt) [inline apps/web/app/api/v1/client/[environmentId]/app/sync/lib/utils.ts:6]: I would assume that a `check...' method would return true or false, but this function modifies the survey and returns the modified one. So maybe something like `replaceAttributeRecalls()` would be more appropriate?
(still not sold on it, but maybe in that direction)

**Review by mattinannt — APPROVED** [review]
@Dhruwang thanks for the PR :-) Looks great! 💪

I like the approach of referencing attributes especially since we are now moving into a more person-centric direction.
I would also love to have this feature for link surveys and will write a ticket for it :-)

**Pull request merged**